server/etcdserver/api/membership/cluster.go

```diff
@@ -303,9 +303,27 @@ func (c *RaftCluster) Recover(onSet func(*zap.Logger, *semver.Version)) {
 
 // ValidateConfigurationChange takes a proposed ConfChange and
 // ensures that it is still valid.
-func (c *RaftCluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
+func (c *RaftCluster) ValidateConfigurationChange(cc raftpb.ConfChange, shouldApplyV3 ShouldApplyV3) error {
 	// TODO: this must be switched to backend as well.
-	membersMap, removedMap := membersFromStore(c.lg, c.v2store)
+	var membersMap map[types.ID]*Member
+	var removedMap map[types.ID]bool
+	if c.v2store != nil {
+		membersMap, removedMap = membersFromStore(c.lg, c.v2store)
+		if err := c.validateConfigurationChange(cc, membersMap, removedMap); err != nil {
+			return err
+		}
+	}
+
+	if c.be != nil && shouldApplyV3 {
```

Comment:
`shouldApplyV3` is set to `true` only when the applying entry's index is greater than consistent index for avoiding applying EntryConfChange entriy twice , say removing a etcd member twice on v3 backend which unnecessarily increases the index. ref. https://github.com/etcd-io/etcd/pull/12854

So, this configuration change validation does not need to be gated by `shouldApplyV3`. Did I understand it correctly? I think it's harmless but just not necessary.

Reply:
True, should have no side effects. We will just validate v3 unnecessarily. 

On the other hand, I like the symmetry between validation other normal apply logic. It should make merging those paths easier when we will stop restoring from v2 snapshot.

Reply:
shouldApplyV3 is needed here because when cluster recovers without a snapshot and/or backend is ahead, the validation will be based on 'future' information and hence not correct. Similar to AddMember and related operations that make the change, the validation should also refer to backend information only if it is not in future.

Reply:
That's scary. @serathius 
I'm afraid we this change we are getting to be hit by linearizability problems. 

Let's consider situation when: 
  1. Cluster is doing some operations on members, where `foo` already: add(foo), remove(foo), 
  2. In healthy situation the add(foo) fails on the all members and remove(foo) succeeds on all the members. 
  3. Now one of this members crashes before snapshot. The member manages to commit remove(foo) to it's backend.  
  4. The member starts recover from snaoshit & reply scenario
  5. It will on the `add(foo)`:
  a) skip the validation (and even if not skipping, the validation would pass as foo is already removed in HEAD)
  b) apply the add(foo) modification to RAFT  -> (while the other members didn't applied it).  
And this leads to a state when one RAFT member is different than the others...

Now the fought experiment: 
  Why this is a problem for confChange and not for Txns that could also fail to compute precondition based on the HEAD ?  Because we skip the applied transaction entirely and trust the backend has proper state ? 

How we can solve this problem: 
 1. We should have @serathius linearability tests on config changes before we will modify the code, as it's very subtle. 
 2. The ultimate goal should be that we don't reply ANY log entries before the 'backend' state. 
 We should think that backend state is the snapshot we restore from, we bootstrap RAFT and membership store... and only apply the future logs.  
 3. In such situation we restore membership information and configure RAFT solely on the state of backend, so we don't need to apply any 'previous' conf changes onto RAFT at all. 

I'm afraid such change needs to happen atomically: 
 - either we restore from any ~V2 snapshot, and incrementally follow conf-changes based on V2 state. 
 - or we only restore from backend virtual snapshot (ignoring snapshots in the RAFT log ?) and configure RAFT based on the backend state. 

Potentially we might try to do some reconciliation:... buffer all the config changes during the reply without setting them in RAFT... and when we reach state of the 'backend', we would reconcile current config in RAFT with desired config in BACKEND. 
But it seems error prone and definitely would require a lot of robustness-testing.

Reply:
@ptabor how about this solution - now the PR has a replay store which should be consulted if backend cannot be consulted. Can you please take a look?

Reply:
pinging on this one. Once the changes to create v2 snapshot from v3 are done, I would like to remove the store from RaftCluster. This code path for validation will block that change. Please let me know if the proposed solution as seen in the diff looks ok. @ptabor @serathius @ahrtr

Reply:
Sorry for the late response. I just get time to look into this today.

> 
> Let's consider situation when:
> 
> 1. Cluster is doing some operations on members, where foo already: add(foo), remove(foo),
> 2. In healthy situation the add(foo) fails on the all members and remove(foo) succeeds on all the members.
> 3. Now one of this members crashes before snapshot. The member manages to commit remove(foo) to it's backend.
> 4. The member starts recover from snaoshit & reply scenario
> 5. It will on the add(foo):
>  a) skip the validation (and even if not skipping, the validation would pass as foo is already removed in HEAD)
>  b) apply the add(foo) modification to RAFT -> (while the other members didn't applied it).
>  And this leads to a state when one RAFT member is different than the others...

At step 5, if it replays the `add(foo)`, it will also replay `remove(foo)`; eventually the state is also correct. Note intermediate inconsistent isn't a problem, because the member can't serve linearilizablity request if its data isn't in sync with the leader. @ptabor 

> if c.be != nil && shouldApplyV3 {
>     ......
> } else {
>     // line 325 ~ 327
>     membersMap, removedMap = c.rs.Members()
>     c.validateConfigurationChange(...)
>     ...
> }
> 

The else branch (line 325 ~ 327) seems not correct. If I understand your change correctly, the replayStore (c.rs) only contains new configure changes after the member gets started. In other words, it may not contain the complete membership data. 

I also do not understand why do you add a replayStore.

Reply:
After reading through all the etcd membership changes (don't ask how long it took) I think I understand the problem (somewhat). High level the problem is not necessarily with validation, but which parts of RaftCluster state we treat as the "current". Note there are 3 states: cache (fields in RaftCluster struct, V2 store and V3 store). I will resolve the comment as the discussion diverged from the original problem and leave new comments on the code directly.